llama_stack/providers/inline/inference/vllm/vllm.py

```diff
@@ -4,238 +4,849 @@
+        resolved_llama_model = llama_models.sku_list.resolve_model(
+            model.provider_model_id
+        )
+        if resolved_llama_model is not None:
+            # Load from Hugging Face repo into default local cache dir
+            resolved_model_id = resolved_llama_model.huggingface_repo
+
+            # Detect a geniune Meta Llama model to trigger Meta-specific preprocessing.
+            # Don't set self.is_meta_llama_model until we actually load the model.
+            is_meta_llama_model = True
+        else:  # if resolved_llama_model is None
+            # Not a Llama model name. Pass the model id through to vLLM's loader
+            resolved_model_id = model.provider_model_id
+            is_meta_llama_model = False
+
+        logger.info(f"Model id {model} resolved to {resolved_model_id}")
+
+        if self.resolved_model_id is not None:
+            if resolved_model_id != self.resolved_model_id:
+                raise ValueError(
+                    f"Attempted to serve two LLMs (ids '{self.resolved_model_id}') and "
+                    f"'{resolved_model_id}') from one copy of provider '{self}'. Use multiple "
+                    f"copies of the provider instead."
+                )
+            else:
+                # Model already loaded
+                self.model_ids.add(model.model_id)
+                return model
+
+        if is_meta_llama_model:
+            logger.info(f"Model {resolved_model_id} is a Meta Llama model.")
+        self.is_meta_llama_model = is_meta_llama_model
+        logger.info(f"Preloading model: {resolved_model_id}")
+
+        # If we get here, this is the first time registering a model.
+        # Preload so that the first inference request won't time out.
         engine_args = AsyncEngineArgs(
-            model=model.huggingface_repo,
-            tokenizer=model.huggingface_repo,
+            model=resolved_model_id,
+            tokenizer=resolved_model_id,
             tensor_parallel_size=self.config.tensor_parallel_size,
             enforce_eager=self.config.enforce_eager,
             gpu_memory_utilization=self.config.gpu_memory_utilization,
-            guided_decoding_backend="lm-format-enforcer",
+            max_num_seqs=self.config.max_num_seqs,
+            max_model_len=self.config.max_model_len,
         )
-
         self.engine = AsyncLLMEngine.from_engine_args(engine_args)
 
-    async def shutdown(self):
-        """Shut down the vLLM inference adapter."""
-        log.info("Shutting down vLLM inference provider.")
-        if self.engine:
-            self.engine.shutdown_background_loop()
+        # vLLM currently requires the user to specify the tool parser manually. To choose a tool
+        # parser, we need to determine what model architecture is being used. For now, we infer
+        # that information from what config class the model uses.
+        low_level_model_config = self.engine.engine.get_model_config()
```

Comment:
What's the difference between `self.engine.engine.get_model_config()` and `self.engine.get_model_config()`? This is a bit confusing.

Reply:
I can see how this might be confusing. This line of code drills down through multiple layers of adapters to get at the Hugging Face Transformers tokenizer object buried deep inside the `AsyncLLMEngine` object. The specific path is AsyncLLMEngine -> LLMEngine -> TokenizerGroup -> TokenizerBase.

I've added an explanatory comment to this line and broken the pointer-chasing operations into three steps.